File: leetcode91_Decode Ways_动态规划_(AC).cpp
class Solution {
public:
    int numDecodings(string s) {
    	if(s[0] == '0' || s.length() == 0) return 0;
    	int p1=1,p2=1;
    	for(int i=1;i<s.length();i++){
    		if(s[i] == '0') p1 = 0;
    		if(s[i-1] == '1' || s[i-1] == '2' && s[i] <= '6'){
    			p1 = p2 + p1;
    			p2 = p1 - p2;
    		}
    		else p2 = p1;
    	}
        return p1;
    }
};

File: leetcode91_Decode Ways_动态规划_(AC).py
class Solution(object):
    def numDecodings(self, s):
        """
        :type s: str
        :rtype: int
        """
        if len(s) is 0:
        	return 0
        dp = [1] + [0]*len(s)
        for i in range(1,len(s)+1):
        	dp[i] = 0 if s[i-1]=='0' else dp[i-1]
        	dp[i] += dp[i-2] if (i>1 and (s[i-2]=='1' or (s[i-2] == '2' and s[i-1] <= '6'))) else 0
        return dp[len(s)]